tests/gtest/tcp/tcp_sockopt.cc

```diff
@@ -0,0 +1,324 @@
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <fstream>
+#include <limits>
+#include <stdexcept>
+#include <tuple>
+
+#include "common/def.h"
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+extern struct gtest_configure_t gtest_conf;
+
+struct reusable_cleanable_test_socket {
+  int m_fd;
+  reusable_cleanable_test_socket(int domain, int type, int protocol) {
+    m_fd = socket(domain, type, protocol);
+    EXPECT_GE(m_fd, 0) << "Unable to open the socket";
+    int reuse = 1;
+    auto result =
+        setsockopt(m_fd, SOL_SOCKET, SO_REUSEPORT, &reuse, sizeof(reuse));
+    EXPECT_EQ(result, 0) << "setsockopt failed to set reuse port";
+    result = setsockopt(m_fd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse));
+    EXPECT_EQ(result, 0) << "setsockopt failed to set reuse addr";
+
+    struct timeval tv;
+    tv.tv_sec = 5;
+    tv.tv_usec = 0;
+    setsockopt(m_fd, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
+  }
+  explicit reusable_cleanable_test_socket(int fd) : m_fd{fd} {
+    EXPECT_GE(m_fd, 0) << "Unable to open the socket";
+  };
+  ~reusable_cleanable_test_socket() { close(m_fd); }
+
+  operator int() const { return m_fd; }
+};
+
+struct ipc {
+  enum FifoDirection : size_t { ReadSide, WriteSide };
+  int m_pipe[2];
+  ipc() : m_pipe{-1, -1} {}
+
+  ~ipc() { reset(); }
+  void create() {
+    if (pipe(m_pipe) != 0) {
+      throw std::runtime_error("Pipe not created");
+    }
+  }
+  void reset() {
+    if (m_pipe[ReadSide] != -1) {
+      close(m_pipe[ReadSide]);
+    }
+    if (m_pipe[WriteSide] != -1) {
+      close(m_pipe[WriteSide]);
+    }
+  }
+  bool wait_peer() {
+    if (m_pipe[ReadSide] == -1) {
+      return false;
+    }
+
+    if (m_pipe[WriteSide] != -1) {
+      if (close(m_pipe[WriteSide]) != 0) {
+        return false;
+      }
+      m_pipe[WriteSide] = -1;
+    }
+
+    char buffer[16];
+    auto result = read(m_pipe[ReadSide], buffer, 1) == 1;
+    return result;
+  }
+
+  bool signal_to_peer() {
+    if (m_pipe[WriteSide] == -1) {
+      return false;
+    }
+
+    if (m_pipe[ReadSide] != -1) {
+      if (close(m_pipe[ReadSide]) != 0) {
+        return false;
+      }
+      m_pipe[ReadSide] = -1;
+    }
+
+    return write(m_pipe[WriteSide], "X", 1) == 1;
+  }
+};
+
+using sockopt_parameters = std::tuple<int, int, int, int>;
```

Comment:
Please consider adding test description as in referenced https://github.com/Mellanox/libvma/blob/master/tests/gtest/tcp/tcp_connect.cc#L138